kotest-assertions/kotest-assertions-core/src/commonMain/kotlin/io/kotest/matchers/sequences/matchers.kt

```diff
@@ -79,17 +81,25 @@ fun <T> containExactly(vararg expected: T): Matcher<Sequence<T>?> = containExact
 
 /** Assert that a sequence contains exactly the given values and nothing else, in order. */
 fun <T, C : Sequence<T>> containExactly(expected: C): Matcher<C?> = neverNullMatcher { value ->
-   var passed: Boolean = value.count() == expected.count()
+   val actualIterator = value.withIndex().iterator()
+   val expectedIterator = expected.withIndex().iterator()
+
+   var passed = true
    var failMessage = "Sequence should contain exactly $expected but was $value"
-   if (passed) {
-      val diff = value.zip(expected) { a, b -> Triple(a, b, a == b) }.withIndex().find { !it.value.third }
-      if (diff != null) {
+   while (passed && actualIterator.hasNext() && expectedIterator.hasNext()) {
+      val actualElement = actualIterator.next()
+      val expectedElement = expectedIterator.next()
+      if (eq(actualElement.value, expectedElement.value) != null) {
+         failMessage += " (expected ${expectedElement.value.show().value} at ${expectedElement.index} but found ${actualElement.value.show().value})"
          passed = false
-         failMessage += " (expected ${diff.value.second} at ${diff.index} but found ${diff.value.first})"
       }
-   } else {
-      failMessage += " (expected ${expected.count()} elements but found ${value.count()})"
    }
+
+   if (actualIterator.hasNext() || expectedIterator.hasNext()) {
```

Comment:
We could split this into two if statements and then say "actual had more elements" or "expected had more elements"

Reply:
done